packages/next/next-server/lib/router/router.ts

```diff
@@ -338,6 +338,36 @@ function prepareUrlAs(router: NextRouter, url: Url, as?: Url) {
   }
 }
 
+// Focus a user's attention to the beginning of new content. This is done by focusing the first h1 tag. If an h1 does
+// not exist, focus the main semantic tag. If this does not exist, focus the cursor on the body. This methodology is
+// inspired by Marcy Sutton's accessible client routing user testing. More information can be found here:
+// https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/
+function a11ySmartFocus() {
```

Comment:
So I did not test it yet, but I am pretty sure this is going to announce the document’s title twice. Once from `RouteAnnouncer`, since it’s an assertive live region, and then a second time when focusing the document’s main title. 

I’m not entirely certain, but it feels like you are mixing two approaches here:

1. **Focusing the main heading.** Upon route change, and provided you can guarantee there will be a `<h1>` element regardless of state such as loading, move the focus to the main heading. As a result, its content will be announced.

2. **Announcing the page title.** If the presence of a `<h1>` cannot be guaranteed for any reason (because left to the implementor who might not know, or because depending on loading states for instance), move the focus to a visually hidden node as `<body>`’s first child containing the new document title (which doesn‘t have to be a region whatsoever).

Both are equally good as far as I know. The main difference is where the focus ends up being. In the first approach, the focus is moved to the main page title, which could be anywhere in the page, really. Some sites have it in the title, some have it in the main container, some have it after the navigation, some before… In the second case, the focus is consistently moved at the top of the body (where the `RouteAnnouncer` DOM node lives), just like a regular `href` link would after a page load.

Because you appear to be essentially doing both, I am pretty certain this will announce the document’s main title twice, which is probably not what you want. I might be entirely mistaken though, so feel free to correct me. :)

**Edit:** a little more information in that article, https://kittygiraudel.com/2020/01/15/accessible-title-in-a-single-page-react-application/.

**Edit 2:** Finding the first `<h1>` element on the page could lead to issues if a hidden dialog with a `<h1>` element is being rendered high up the body element. That heading will be queried because it’s the first, but attempting to focus it would not work because it is within a hidden dialog.

Reply:
I agree with the last point about the h1 selector which is a bit dangerous because it contains an assumption about the website structure.

Also the explanation about how the title is now announced twice sounds reasonable to me, but haven't tried what happens.

Reply:
Hey! Thank you so much for taking the time to look at this, and more importantly, for the feedback! ❤️

I think your judgement is 100% valid. The intent here was to bring focus to the most important part of the page; however, this makes quite a few assumptions about the user's site. For example, the heading might be visually hidden. This alone makes me think that while this might be helpful for many sites, it probably should not be added at this level.

The following is probably a moot point, but because the `<RouteAnnouncer />` uses `aria-live="assertive"`, the screen reader is expected to stop what it's doing and announce the new page name. Since the focus & route announcement happen at the same time, most screen readers will give priority to the `<RouteAnnouncer />` and not announce the focused element. I've tested this on a few screen readers.

Look out for a commit soon with this change 😉

Reply:
Oh, I absolutely love this idea! It's _really_ elegant. I can already tell that this will decrease the complexity significantly. Thanks @KittyGiraudel!

Reply:
I don't know if it's solvable here, but I don't see that being discussed—this refocus behavior breaks `autoFocus` elements as they won't be focused on page transition anymore. However the `autoFocus` attribute itself might be a bad thing for a11y, so I'm not sure what the best solution is.

Reply:
@shuding, that's a very good point. `autoFocus` is broken with the current implementation of this PR as well as with the awesome suggestion @kittyGiraudel made. In the name of not breaking any functionality, I'll hold off on the focusing-side of this PR and dedicate a new PR for it in the near-future.

Reply:
Thank you @kyleboss, I'll land the Route Announcer part then 👍